### PR TITLE
chore: Make past event error message user friendly

### LINKED ISSUE #1854: Show user friendly error message when user tries to publish a past event
**Describe the bug**
Show user friendly error message when user tries to publish a past event.

**Would you like to work on the issue?**
Open to all.

### PR DESCRIPTION
**Pull request by ShridharGoel**
Fixes #1854 

Changes: 

Now, user friendly error message would be shown when user tries to publish a past event.

### CODE REVIEW

File: app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java [hunk 1]
```diff
@@ -119,7 +121,17 @@ private void listenChanges() {
     }
 
     public void confirmToggle() {
-        if (Event.STATE_PUBLISHED.equals(event.state)) {
+        boolean isPastEvent = false;
+        try {
+            isPastEvent = DateService.getEventStatus(event).equalsIgnoreCase("past");
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: BTW, this is a horrible logic for checking that the event is past.

**Comment by ShridharGoel** [comment]
Actually, on checking again, found that it is showing an error. Not sure how it was getting published when tried during the demo in the meeting.

Publish past events

**Comment by iamareebjamal** [comment]
Because this check was added 3 days ago on server

**Comment by iamareebjamal** [comment]
Make this error user friendly.

**Comment by ShridharGoel** [comment]
Oh okay. Although I remember that this error message used to display earlier too. The check on the server might have got removed after that.

**Comment by ShridharGoel** [comment]
> Make this error user friendly.

Okay

**Comment by ShridharGoel** [comment]
@iamareebjamal Can you tell how is the current error being displayed? I cannot find anything related to showing of error when trying to toggle state.

**Comment by iamareebjamal** [comment]
https://github.com/fossasia/open-event-organizer-android/blob/development/app/src/main/java/com/eventyay/organizer/utils/ErrorUtils.java

**Comment by ShridharGoel** [comment]
> https://github.com/fossasia/open-event-organizer-android/blob/development/app/src/main/java/com/eventyay/organizer/utils/ErrorUtils.java

Yes but a method from this class needs to be called, right? But it's not there in the toggleState() method.

**Comment by iamareebjamal** [comment]
It'll be in onError of the subscription

**Comment by ShridharGoel** [comment]
Updated. Although please tell if there is some better way of making the error user friendly without implementing the check on the app.

**Comment by iamareebjamal** [comment]
You don't have to add check in the app. You have to parse the error coming from backend

**Comment by ShridharGoel** [comment]
But error coming from backend just contains `starts-at should be after current date-time`.

**Comment by ShridharGoel** [comment]
Also, please check the latest changes in this PR.

**Comment by iamareebjamal** [comment]
You have to parse it in the viewmodel, not here

**Comment by iamareebjamal** [comment]
> But error coming from backend just contains starts-at should be after current date-time.

If the pointer is at starts-at, you know it's date error

**Comment by ShridharGoel** [comment]
> You have to parse it in the viewmodel, not here

But the error is being handled through `progressiveErroneous()` in `ViewTransformers`.

**Comment by iamareebjamal** [comment]
You have the capability to add custom logic in onError of the observable

**Comment by ShridharGoel** [comment]
But onError of the observable only throws `retrofit2.adapter.rxjava2.HttpException: HTTP 422 UNPROCESSABLE ENTITY`

**Comment by iamareebjamal** [comment]
Exactly. There you have it

**Comment by ShridharGoel** [comment]
But HTTP 422 can be due to various reasons, how can we get the `starts-at` pointer from this?

**Comment by iamareebjamal** [comment]
By parsing the body

**Comment by iamareebjamal** [comment]
@liveHarshit Can you please collaborate here?

**Comment by liveHarshit** [comment]
@ShridharGoel Can you please share header and response screenshot for this API call from chrome dev tools?

**Comment by ShridharGoel** [comment]
Headers

Response

**Comment by ShridharGoel** [comment]
Actually, body of `throwable` in onError of the observable remains empty, I don't know why.

**Comment by liveHarshit** [comment]
> Actually, body of `throwable` in onError of the observable remains empty, I don't know why.

Have you tried by clearing application data. And check the event timings?

**Comment by liveHarshit** [comment]
Earlier I was facing the same problem on the frontend, after clearing browser data, it worked.

**Comment by ShridharGoel** [comment]
> And check the event timings?

Event timings are of the past, that's why the error is being shown.

**Comment by liveHarshit** [comment]
> > And check the event timings?
> 
> Event timings are of the past, that's why the error is being shown.

Then it should be a server-side issue

**Comment by iamareebjamal** [comment]
Error should be shown. You just have to parse the JSON body and check the pointer. It's not that hard!

**Comment by iamareebjamal** [comment]
> Actually, body of throwable in onError of the observable remains empty, I don't know why.

How are you accessing the body?

**Comment by ShridharGoel** [comment]
> How are you accessing the body?

if (throwable instanceof HttpException) {
          ResponseBody responseBody = ((HttpException) throwable).response().errorBody();
          JSONObject jsonObject = new JSONObject(responseBody.string());
}

**Comment by iamareebjamal** [comment]
Why not just using https://github.com/fossasia/open-event-organizer-android/blob/development/app/src/main/java/com/eventyay/organizer/utils/ErrorUtils.java#L87

**Comment by ShridharGoel** [comment]
> Why not just using https://github.com/fossasia/open-event-organizer-android/blob/development/app/src/main/java/com/eventyay/organizer/utils/ErrorUtils.java#L87

Yes I had tried using that:

```
com.eventyay.organizer E/ErrorUtils: org.json.JSONException: End of input at character 0 of 
        at org.json.JSONTokener.syntaxError(JSONTokener.java:449)
        at org.json.JSONTokener.nextValue(JSONTokener.java:97)
        at org.json.JSONObject.<init>(JSONObject.java:156)
        at org.json.JSONObject.<init>(JSONObject.java:173)
        at com.eventyay.organizer.utils.ErrorUtils.getErrorTitleAndDetails(ErrorUtils.java:93)
        at com.eventyay.organizer.core.event.dashboard.EventDashboardPresenter.lambda$toggleState$6$EventDashboardPresenter(EventDashboardPresenter.java:150)
        at com.eventyay.organizer.core.event.dashboard.-$$Lambda$EventDashboardPresenter$VB8dX52hCyxTmE3xyoX4KpXi9ms.accept(lambda)
        at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
        at io.reactivex.internal.operators.observable.ObservableDoFinally$DoFinallyObserver.onError(ObservableDoFinally.java:85)
        at io.reactivex.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onError(ObservableDoOnEach.java:117)
        at io.reactivex.internal.operators.observable.ObservableDoFinally$DoFinallyObserver.onError(ObservableDoFinally.java:85)
        at io.reactivex.internal.observers.DisposableLambdaObserver.onError(DisposableLambdaObserver.java:65)
        at io.reactivex.internal.observers.DisposableLambdaObserver.onError(DisposableLambdaObserver.java:65)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.checkTerminated(ObservableObserveOn.java:281)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:172)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
        at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:124)
        at android.os.Handler.handleCallback(Handler.java:754)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:163)
        at android.app.ActivityThread.main(ActivityThread.java:6205)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:794)
```

**Comment by iamareebjamal** [comment]
Ensure it is not being called twice

**Comment by ShridharGoel** [comment]
> Ensure it is not being called twice

Yes, the problem was because of this. Updated.

File: app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java [hunk 2]
```diff
@@ -144,7 +145,13 @@ public void toggleState() {
                         getView().onSuccess(utilModel.getResourceString(R.string.draft_success));
                     }
                 },
-                throwable -> event.state = Event.STATE_DRAFT.equals(event.state) ? Event.STATE_PUBLISHED : Event.STATE_DRAFT);
+                throwable -> {
+                    if (ErrorUtils.getErrorDetails(throwable).toString().contains("starts-at should be after current date-time"))
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: Check the source of error, not the entire string

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: Ok, also the tests are failing. Can you please guide why? I think the tests should not have required any changes.

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: > Check the source of error, not the entire string

Updated.

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: https://circleci.com/gh/fossasia/open-event-organizer-android/3385?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: https://3385-65606855-gh.circle-artifacts.com/0/reports/tests/testPlayStoreDebugUnitTest/index.html

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/event/dashboard/EventDashboardPresenter.java]: @iamareebjamal Any help on why the tests aren't passing? I have discussed with @anhanh11001 but he is also not able to figure out the reason.

**Comment by iamareebjamal** [comment]
Because now it won't throw the error as expected in the test. It is conditionally thrown AFAIK

**Comment by ShridharGoel** [comment]
> Because now it won't throw the error as expected in the test. It is conditionally thrown AFAIK

But the `else` block will throw the error I think.

**Comment by iamareebjamal** [comment]
Use debugger

**Pull request closed without merging**